### PR TITLE
[ios] store OTA updates in Application Support directory

### PR DESCRIPTION
**Pull request by esamelson**
# Why

A few users have reported issues stemming from iOS OTA updates being purged from the cache directory by iOS. I've fixed all of the individual issues/cases that I'm aware of, but a more permanent solution is to move OTA updates to a more persistent location on disk.

However, this does not come without some risk. I'm opening this PR for discussion of a few possibilities.

# How

~1. In addition to storing everything in `NSCachesDirectory` we define an `NSURLCache` over the JS bundle source directory. This comes with a max size of 20 MB (which we define) and can means it can be cleared by the OS. The first commit on this branch removes the `NSURLCache` for this folder in standalone apps.~

2. The ~second~ first commit on this branch moves all cached files to the `NSDocumentDirectory`. However, this can possibly raise some flags with Apple -- according to their [data storage guidelines](https://developer.apple.com/icloud/documentation/data-storage/index.html), only user-generated data should be stored in the documents directory. There are a few reports of apps getting rejected for this reason on [stack overflow](https://stackoverflow.com/questions/6879860/when-are-files-from-nscachesdirectory-removed/8830746). We can get around this by setting the "exclude from backups" key on all the cache directories we create, but it seems risky to expose every Expo app to this extra scrutiny.

3. The final commit on this branch adds some basic reaping to shell apps by adding SDK version subfolders in all of the caches and trying to remove folders that are for an older SDK version.

@ide @brentvatne curious to hear your thoughts on this -- I'm leaning towards saying it's too risky for now.

# Test Plan

I've done some very basic testing in the Expo client to make sure that OTA updates are downloaded and loaded properly with these changes and that the reaping works as expected. Also tested a shell app to make sure it cached, loaded and reaped properly, and correctly fell back to downloading the bundle instead of showing an error screen when the disk was in a bad state (partially reaped).

### CODE REVIEW

**Comment by ide** [comment]
I agree with leaning towards omitting step 2. With SQLite I think the document directory is less of a risk because SQLite appears to be less of a cache artifact and more of a persistent database.

What is the purpose of step 1? It seems to me that it's an extra caching layer but doesn't make our JS files more permanent.

And for step 3 (the reaper), is your thinking that by having a reaper, the OS is less likely to evict files on its own?

**Comment by esamelson** [comment]
Hmm, reading into `NSURLCache` more I think I misunderstood how it works and you're first, that first commit wouldn't actually achieve anything worthwhile here. So the question is, do we want to take this chance of moving things into the Documents directory and potentially opening up more red flags during app review. Seems like the answer is probably no for now.

I suppose we could include reaping by SDK version in the caches directory, to make the OS less likely to evict files we want to keep. It's not the full-fledged solution for this issue that we were looking for, though.

We'll need to think about this more carefully for the new updates module -- agree that a SQLite file is less risky, but in the current implementation we store all of the assets (including JS bundles) on disk separate from the SQLite file. We'll probably need to move them to the caches directory and carefully handle any cases where assets were deleted from disk.

**Comment by ide** [comment]
Reading https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html it sounds like the Library directory is maybe the most appropriate location to use. It is backed up (whether we want to enable that is a question to answer) and it is not exposed to the user, which is what we want. The documents directory is exposed to the user, but we definitely do not want that.

File: ios/Exponent/Kernel/AppLoader/CachedResource/EXCachedResource.h [hunk 1]
```diff
@@ -43,6 +43,13 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (BOOL)removeCache;
 
+/**
+ *  Creates cache directory with the given name and returns path.
```
Comment (ide) [inline ios/Exponent/Kernel/AppLoader/CachedResource/EXCachedResource.h]: need to update this comment!

**Review by ide — APPROVED** [review]

File: ios/Exponent/Kernel/AppLoader/CachedResource/EXCachedResource.h (hunk 1 above)
Comment (esamelson) [reply ios/Exponent/Kernel/AppLoader/CachedResource/EXCachedResource.h]: ah yes, thanks for the catch!

**Pull request merged**